Fix caching of the degree of a transformation

james-d-mitchell opened the pull request:
This pull request fixes the caching of the degree of a transformation.
Previously, although there was space reserved in a transformation for
caching the degree, it was in fact never cached. This caused a problem
in `HASH_FUNC_FOR_TRANS` where it was assumed that the degree was cached
after it was calculated. In particular, 4 bytes transformations were
trimmed using the (non-)cached value of the degree, which was 0, and
this resulted in non-identity transformations having their value changed
to the identity transformation when they were hashed.

For example, without this pull request:

     gap> x := Transformation([1,1]) ^ (1,2)(3,70000);
    Transformation( [ 2, 2 ] )
    gap> IsTrans4Rep(x);
    true
    gap> HASH_FUNC_FOR_TRANS(x, 101);
    22
    gap> x;
    IdentityTransformation

and with:

    gap> x := Transformation([1,1]) ^ (1,2)(3,70000);
    Transformation( [ 2, 2 ] )
    gap> IsTrans4Rep(x);
    true
    gap> HASH_FUNC_FOR_TRANS(x, 101);
    41
    gap> x;
    Transformation( [ 2, 2 ] )

markuspf commented:
Could you provide a minimal (previously) failing example for testing purposes? I assume this requires ORB hashes to occur?

markuspf commented:
This patch fixes the described problem, and looks correct to me.

A minor nitpick is that it conflates a lot of whitespace changes with the actual fix for the issue, so it is hard to spot which part of the patch is actually relevant.

We will suggest making this a comment in `CONTRIBUTING.md` and a requirement for future pull requests.

Maybe there should also be a test in `trans.tst` for this?

james-d-mitchell commented:
Yeah, sorry about the whitespace changes. I thought these changes were so minor that it wouldn't matter in this case. I cleaned up the code, and fixed it at the same time. I wasn't sure if making pull requests with changes to whitespace only was appropriate. Is it? I will add the tests shortly.

markuspf commented:
I suggest to refrain from whitespace changes (for extensive reformatting) for now.

We have planned to agree to a uniform code-style for GAP library and C kernel code at the January meeting. We will then make a couple of commits that reformat the code according to the agreed style.

It is perfectly acceptable in my opinion to have whitespace cleanup/reformatting commits if they are clearly marked as that (maybe that should also be in the guidelines for contributing).

fingolfin commented:
IMHO, the most important reason why one should not fix whitespace / indention changes with other changes is that they quickly obscure what's going on. I have a very hard time figuring out what you actually changed, and hence it's harder to review the change for correctness. Hence I am less likely to even try doing it...

alex-konovalov commented:
A practical hint:
* don't use editor which automatically changes whitespace / indention 
* use `git diff` to review changes before commits, so you will be able to spot such things happening.

james-d-mitchell commented:
Thanks for the hints @alex-konovalov but I know precisely what my text editor is doing, and the whitespace changes in this PR were intentional.

Anyway, I have closed this PR and will make a new one without any whitespace changes.
